x/feedist/handler_test.go

```diff
@@ -0,0 +1,551 @@
+// call.
+func asSeqID(i int64) []byte {
+	b := make([]byte, 8)
+	binary.BigEndian.PutUint64(b, uint64(i))
+	return b
+}
+
+func coinp(w, f int64, ticker string) *x.Coin {
+	c := x.NewCoin(w, f, ticker)
+	return &c
+}
+
+func TestFindGdc(t *testing.T) {
+	cases := map[string]struct {
+		want   int32
+		values []int32
+	}{
+		"empty": {
+			want:   0,
+			values: nil,
+		},
+		"one element": {
+			want:   7,
+			values: []int32{7},
+		},
+		"two elements": {
+			want:   3,
+			values: []int32{9, 6},
+		},
+		"three elements": {
+			want:   3,
+			values: []int32{9, 3, 6},
+		},
+		"four elements": {
+			want:   6,
+			values: []int32{12, 6, 18},
+		},
+		"less common divisors": {
+			want:   2,
+			values: []int32{24, 12, 64, 18},
+		},
+		"prime numbers": {
+			want:   1,
+			values: []int32{67, 71, 73, 79, 83, 89, 97},
+		},
+	}
+
+	for testName, tc := range cases {
+		t.Run(testName, func(t *testing.T) {
+			got := findGcd(tc.values...)
+			if got != tc.want {
+				t.Fatalf("want %d, got %d", tc.want, got)
+			}
+		})
+	}
+}
+
+func TestDistribute(t *testing.T) {
+	cases := map[string]struct {
+		recipients []*Recipient
+		ctrl       *testController
+		// Each MoveCoins call on the testController result in creation
+		// of a movecall. Those can be used later to validate that
+		// certain MoveCoins calls were made.
+		wantMoves []movecall
+		wantErr   error
+	}{
+		"zero funds is not distributed": {
+			recipients: []*Recipient{
+				{Address: weave.Address("address-1"), Weight: 1},
+				{Address: weave.Address("address-2"), Weight: 2},
+			},
+			ctrl: &testController{
+				balance: nil,
+			},
+			wantErr: nil,
+		},
+		"tiny funds are not distributed if cannot be split": {
+			recipients: []*Recipient{
+				{Address: weave.Address("address-1"), Weight: 1},
+				{Address: weave.Address("address-2"), Weight: 2},
+			},
+			ctrl: &testController{
+				balance: x.Coins{coinp(0, 1, "ETH")},
+			},
+			wantErr: nil,
+		},
+		"simple distribute case": {
+			recipients: []*Recipient{
+				{Address: weave.Address("address-1"), Weight: 1},
+				{Address: weave.Address("address-2"), Weight: 2},
+			},
+			ctrl: &testController{
+				balance: x.Coins{coinp(3, 0, "BTC")},
+			},
+			wantErr: nil,
+			wantMoves: []movecall{
+				{dst: weave.Address("address-1"), amount: x.NewCoin(1, 0, "BTC")},
+				{dst: weave.Address("address-2"), amount: x.NewCoin(2, 0, "BTC")},
+			},
+		},
+		"distribution splits whole into fractional": {
+			recipients: []*Recipient{
+				{Address: weave.Address("address-1"), Weight: 1},
+				{Address: weave.Address("address-2"), Weight: 2},
+			},
+			ctrl: &testController{
+				balance: x.Coins{coinp(1, 0, "BTC")},
+			},
+			wantErr: nil,
+			wantMoves: []movecall{
+				// One cent is left on the revenue account,
+				// because it is too small to divide.
+				{dst: weave.Address("address-1"), amount: x.NewCoin(0, x.FracUnit/3, "BTC")},
+				{dst: weave.Address("address-2"), amount: x.NewCoin(0, x.FracUnit/3*2, "BTC")},
```

Comment:
No problem leaving a rest. I'd rather it was fair always that try to divide the smallest fraction.

Reply:
I don't think this is `666666667` is correct. You are giving one recipient proportionally more money than the other. If we respect the weights, they should receive respectively 33..333 and 66..666. One cent is left because you cannot split it any further. Gifting it to anyone is not fair.